planex/fileupdate.py

```diff
@@ -0,0 +1,62 @@
+"""
+<pipe> | idem <file> overwrites <file> with the output of <pipe>
+if the output of <pipe> is different from the contents of <file>,
+otherwise leaves <file> untouched.
+"""
+
+import hashlib
+import os
+import shutil
+import sys
+import tempfile
+
+def sameSize(infile, outfile):
+    infile.seek(0)
+
+    insize = os.fstat(infile.fileno()).st_size
+    outsize = os.fstat(outfile.fileno()).st_size
+
+    print "infile size:  %d bytes" % insize
+    print "outfile size: %d bytes" % outsize
+
+    return insize == outsize
+
+def sameHash(infile, outfile):
+    infile.seek(0)
+    outfile.seek(0)
+
+    inhash = hashlib.md5()
+    inhash.update(infile.read())
+    outhash = hashlib.md5()
+    outhash.update(outfile.read())
+
+    print "infile hash:  %s" % inhash.hexdigest()
+    print "outfile hash: %s" % outhash.hexdigest()
+
+    return inhash.digest() == outhash.digest()
+
+class FileUpdate:
+    def __init__(self, filename):
+        self.infile = tempfile.TemporaryFile()
+        self.filename = filename
+
+    def __enter__(self):
+        return self.infile
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        try:
+            with open(self.filename, 'rb') as outfile:
+                if (sameSize(self.infile, outfile) and 
+                    sameHash(self.infile, outfile)):
+                    print "No changes"
+                    return
+
+        except IOError as ioe:
+            # Errno == 2 => output file does not exist
+            if ioe.errno != 2:
+                raise
+
+        print "Copying infile to outfile"
+        self.infile.seek(0)
+        with open(self.filename, 'wb') as outfile:
+            shutil.copyfileobj(self.infile, outfile)
```

Comment:
In this case I've used TemporaryFile so I don't have a filename and can't use shutil.move().   If I used TemporaryFile I could shutil.move() the file, but since /tmp is usually a separate filesystem the move would turn into a copy anyway.